+ sanma(triple) zhengma dict

p1uxtar opened the pull request:
### Brief summary of what the package does

This is the dict/thesaurus file of the Chinese character input method Sanma (triple) Zhengma. This scheme is an improved version of [Zhengma](https://zh.wikibooks.org/wiki/%E9%83%91%E7%A0%81%E8%BE%93%E5%85%A5%E6%B3%95), which has been developed by Zhizhi for more than ten years.
This package is a solution of smzm (Sanma Zhengma) adapted to the [pyim](https://github.com/tumashu/pyim) input method, and its mode of action is consistent with [pyim-wbdict](https://github.com/tumashu/pyim-wbdict) and [pyim-cangjie5dict](https://github.com/p1uxtar/pyim-cangjie5dict), etc.

### Direct link to the package repository

https://github.com/p1uxtar/pyim-smzmdict

### Your association with the package

I'm the maintainer and one of the users of this project, the original author (Zhizhi) sent me the dict source and he doesn't use GitHub or other version control software.

### Relevant communications with the upstream package maintainer

The author of pyim [tumashu](https://github.com/tumashu) has agreed that this scheme can be incorporated into pyim. Once melpa is included, he will merge the PR of pyim simultaneously.

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I have confirmed some of these without doing them

riscy commented:
Thanks for this - please double-check that you correctly confirmed everything on the checklist.  Here's a quick first pass --

### pyim-smzmdict.el ###

byte-compile (using Emacs 27.1):
```
pyim-smzmdict.el:40:1:Warning: the function `pyim-extra-dicts-add-dict' is not
    known to be defined.
```

checkdoc (using version 0.6.2):
```
Warning (emacs): 
pyim-smzmdict.el:39: The footer should be: (provide 'pyim-smzmdict)\n;;; pyim-smzmdict.el ends here
```

package-lint (using version 20210227.314):
```
3 issues found:
1:61: warning: You should depend on (emacs "24.1") if you need lexical-binding.
7:13: error: Package URLs should be a single HTTPS or HTTP URL.
8:3: warning: You should include standard keywords: see the variable `finder-known-keywords'.
```


### Loadability ###

Verifying ability to #'load each file:
```
Loading pyim-smzmdict.el
Done.
```

### Package ###

- Add a LICENSE file that GitHub can detect (e.g. no markup) if possible
  See: https://github.com/licensee/licensee
- Add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to pyim-smzmdict.el

p1uxtar commented:
Thank you for your work on this, Chris.

I have checked the issues listed above. 

### pyim-smzmdict.el ###

The `pyim-extra-dicts-add-dict' function is provided by [pyim](https://github.com/tumashu/pyim). Pyim-smzmdict depends on pyim to run correctly, I have tested this feature, and refer to the code of [pyim-cangjie5dict](https://github.com/p1uxtar/pyim-cangjie5dict) and [pyim-wbdict](https://github.com/tumashu/pyim-wbdict). When using pyim to input Chinese characters, pyim-smzmdict provides candidate ones. It can be understood imperfectly as one of the backends of pyim. Thus I have added comments to the Commentary to prompt novice users to install pyim firstly.

I have fixed the warning by modifying the footer and package address to be consistent with the current information of this project.

I used analogous keywords in the previous version which style is the same as the pyim-*dict projects just mentioned. Sorry if this caused you trouble.

### Loadability ###

Similarly, autoload related abilities also need to cooperate with pyim.

### Package ###

I have added the GPL license based on the GitHub template. Zhizhi, who provided the data sources, also has no objection to free sharing and dissemination.


Thanks again!

riscy commented:
Thanks for the update.  I see currently package-lint is erroring because it can't parse the buffer.  This is because the summary line lists an incorrect filename.

> The `pyim-extra-dicts-add-dict' function is provided by pyim.

In that case you should simply `(require 'pyim)`, and checking `(if (featurep 'pyim)` within the enablement code becomes extraneous.  Lastly I notice you check `(locate-library "pyim-smzm.el")` but that's not the name of this package, so I wonder if this check will fail for some users.

p1uxtar commented:
Hello, Chris

I think I need to explain to you again that the pyim should be enabled firstly to run these pyim-*dict packages correctly. They cannot be used alone without pyim, thus they must also be configured in accordance with pyim's README, such as `(setq pyim-default-scheme 'smzm)` is used to enable this scheme, not just `(require 'pyim-smzmdict)`. So I think it is enough to prompt the user to install in the [README](https://github.com/tumashu/pyim/blob/master/README.md), and writing other configurations may conflict or ambiguity with the default configuration of pyim.

`(locate-library "pyim-smzmdict.el")` I had fixed, also some others details.

Thank you again!

riscy commented:
> I think I need to explain to you again that the pyim should be enabled firstly

Fair enough. :)  I will probably need similar reminding in the future.  In this case you can achieve a clean byte-compile by forward-declaring the function in question with `declare-function`.  Emacs gives you a way to check these declarations with `check-declare-file`.

riscy commented:
Part of the confusion may have been because you write "User needs to install Pyim before using this package" but I think it would be clearer to write that the user needs to _require_ or _load_ pyim before loading your package.

p1uxtar commented:
I had run `check-declare-file` and `byte-compile-file` etc. then no warning or other messages emerged, maybe because I had installed and configured `pyim` correctly on my Emacs.

In other words, I still don’t know how to prove to you that this package works correctly. Could you tell me what to do next? For instance, which log or other files to check?

riscy commented:
Yes, most likely you're not seeing all of the compilation warnings because you're compiling your code in an instance of Emacs that has already loaded pyim, or some past version of it at least.  Using `flycheck` helps with this, or you can restart emacs using `emacs -Q` and try compiling from there.

If you add the `declare-function` I mentioned, all users will see a clean byte-compile too.

riscy commented:
This is a minor issue, but if I've been unclear feel free to ping.  I'll mark this awaiting upstream in the meantime.

p1uxtar commented:
I have restart Emacs using `emacs -Q`, then only eval the pyim config and this smzmdict. It works correctly and no messages or warnings emerged after I had run `check-declare-file` and `byte-compile-file`.

So I decided to add the configures adapted to pyim for this project, including `(require 'pyim)` and scheme-add with the prefix `_`, etc. Thus the users (after putting `pyim` and `pyim-smzmdict` into the path emacs can read of course) can enable smzm by two operations: 1) require smzm-dict; 2) eval one function: `(pyim-smzmdict-enable)` then execute `(toggle-input-method)`, and the `pyim` enabled at the same time.

Thank you!

riscy commented:
Sounds good to me.  Thanks for this!
